### PR TITLE
Refactor getQuestionResultStatistics* methods in FeedbackMsqQuestionDetails class #6659

### LINKED ISSUE #6659: FeedbackMsqQuestionDetails class: Refactor getQuestionResultStatistics* methods
The two methods `getQuestionResultStatisticsCsv()` and `getQuestionResultStatisticsHtml()` methods have 80+ lines. These methods can be refactored.

### PR DESCRIPTION
**Pull request by AnaghHegde**
Fixes #6659

**Outline of Solution**

The two methods getQuestionResultStatisticsCsv() and getQuestionResultStatisticsHtml() has many things in common. So i have added three extra methods which is common to both.

### CODE REVIEW

**Comment by samsontmr** [comment]
You can actually set your new local branch to track your old remote branch and push there instead of constantly opening new PRs...

**Comment by AnaghHegde** [comment]
The issue is remote branch is deleted, not the local branch. I tried to find a solution but couldn't find any. If it was just local branch what you have said will work, but in this case i dont think it will work.

**Comment by junhaoyap** [comment]
@AnaghHegde travis build is failing

**Comment by AnaghHegde** [comment]
Oh i didn't see that. I will push a patch.

**Comment by AnaghHegde** [comment]
@samsontmr Ready for review.

**Comment by AnaghHegde** [comment]
@samsontmr What part did i miss ?. I couldn't find anything. You are talking about the use of `containsEmptyResponse` as a instance-level variable ?.

**Comment by samsontmr** [comment]
> You are talking about the use of containsEmptyResponse as a instance-level variable ?.

Yes.

**Comment by AnaghHegde** [comment]
Yes. Because it's used to check if the non empty response is there (from line number 477) and line 441 depends on this value. So either i have to return both (as a string and split them later) or make `numChoicesSelected` as global or the current one.

**Comment by AnaghHegde** [comment]
Ready for review.

**Comment by AnaghHegde** [comment]
With all due respect, I have somethings to ask: 

>Make a single commit with the new method that you extracted.

My first commit does that only except that the commit message is different.

>Make one commit each for each of the two methods (getQuestionResultStatisticsCsv() and getQuestionResultStatisticsHtml()) 

Does it mean that every logical changes done in these two methods should be one single commit ? or different commit?.

**Comment by thenaesh** [comment]
@AnaghHegde Your first commit is basically the whole refactoring. And you should really go have a look at the diff.

What I'm saying is this: In the first commit, there should be a new method, and nothing should have been removed (in other words, the diff should not have any removals in it, and should have only one chunk of additions). Then in the second and third commits, you should update the two methods (csv and html) to make use of this newly defined method. In each of these commit diffs, there should be a section of removals followed by a section of addition, showing clearly what was modified.

This is what is meant by "commit every logical change". It should be very clear exactly what each commit is meant to do. This will not only help reviewing, but also help keep the commit histories clean.

**Comment by AnaghHegde** [comment]
@thenaesh Right now i am having my 3rd internal assement is going on. This week end i will work on it.

**Comment by AnaghHegde** [comment]
@thenaesh I hope i have done as you asked. Ready for review.

**Comment by AnaghHegde** [comment]
@thenaesh Ready for review.

**Comment by samsontmr** [comment]
I see you have managed to organise your commits and avoid making `containsEmptyResponse` an instance level variable. Well done!

**Comment by AnaghHegde** [comment]
@samatdav Ready for review.

**Comment by AnaghHegde** [comment]
@wkurniawan07 Ready for review.

File: src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java [hunk 1]
```diff
@@ -403,9 +403,33 @@ public String getQuestionResultStatisticsHtml(List<FeedbackResponseAttributes> r
...
+        }
+        //Use same template as MCQ for now, until they need to be different.
+        return Templates.populateTemplate(FormTemplates.MCQ_RESULT_STATS, Slots.FRAGMENTS, fragments.toString());
+    }
+
+    /**
+     * Check if there is a non empty response.
+     * @return -1 if there is no empty response else number of empty response.
+     */
+    private int getQuestionResultStatistics(
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java]: The method name does not match its method signature. If it really is about getting statistics, I'd expect some kind of statistics class or object as the return type, not an integer or "-1" if there are no empty responses (what does empty response have to do with it?).

File: src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java [hunk 2]
```diff
@@ -457,21 +473,31 @@ public String getQuestionResultStatisticsHtml(List<FeedbackResponseAttributes> r
...
+        return numChoicesSelected;
+    }
 
-        StringBuilder fragments = new StringBuilder();
-        for (Entry<String, Integer> entry : answerFrequency.entrySet()) {
-            fragments.append(Templates.populateTemplate(FormTemplates.MCQ_RESULT_STATS_OPTIONFRAGMENT,
-                                Slots.MCQ_CHOICE_VALUE, entry.getKey(),
-                                Slots.COUNT, entry.getValue().toString(),
-                                Slots.PERCENTAGE, df.format(100 * (double) entry.getValue() / numChoicesSelected)));
+    private int checkContainsNonEmptyResponse(List<String> answerStrings, Map<String, Integer> answerFrequency) {
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java]: Similarly here. One would not expect a method named `check*` to return an int, no?

Also, if the method is only used once, you might not need to abstract it out. Do not abstract out a method or routine just for the sake of it (e.g. to make the call site look shorter). You need to be able to explain the abstracted out method on its own.

File: src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java (hunk 1 above)
Comment (AnaghHegde) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java]: ok. Shall i change the method name to a meaningful name?

Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java]: Yes, you have to.

**Comment by AnaghHegde** [comment]
@wkurniawan07 Ready for review.

File: src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java [hunk 3]
```diff
@@ -403,9 +403,33 @@ public String getQuestionResultStatisticsHtml(List<FeedbackResponseAttributes> r
             return "";
         }
 
-        boolean isContainsNonEmptyResponse = false; // we will only show stats if there is at least one nonempty response
-
         Map<String, Integer> answerFrequency = new LinkedHashMap<String, Integer>();
+        int numChoicesSelected = getNumberOfEmptyResponse(responses, answerFrequency);
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java]: Here you assign the return value of `getNumberOfEmptyResponse` to a variable called `numChoicesSelected`. Does it not seem odd to you? Either one of the names is inaccurate; could even be both.

File: src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java [hunk 4]
```diff
@@ -436,18 +460,10 @@ public String getQuestionResultStatisticsHtml(List<FeedbackResponseAttributes> r
                 answerStrings.remove(otherAnswer);
             }
 
-            for (String answerString : answerStrings) {
-                if (answerString.isEmpty()) {
-                    continue;
-                }
-
+            int numChoices = getNumberOfEmptyResponseOfQuestion(answerStrings, answerFrequency);
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java]: Similarly here. You're assigning the "number of empty response" to "number of choices".

**Comment by AnaghHegde** [comment]
@wkurniawan07 Ready for review.

File: src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java [hunk 5]
```diff
@@ -403,9 +403,33 @@ public String getQuestionResultStatisticsHtml(List<FeedbackResponseAttributes> r
...
+                                Slots.PERCENTAGE,
+                                df.format(100 * divideOrReturnZero((double) entry.getValue(), numChoicesSelected))));
+
+        }
+        //Use same template as MCQ for now, until they need to be different.
+        return Templates.populateTemplate(FormTemplates.MCQ_RESULT_STATS, Slots.FRAGMENTS, fragments.toString());
+    }
+
+    /**
+     * Check if there is a non empty response.
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java]: The Javadoc still is not in sync with the method name. The method name suggests "getting number of responses", not "checking if there are non-empty responses".

File: src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java (hunk 4 above)
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java]: This change request is not addressed yet.

**Comment by whipermr5** [comment]
@wkurniawan07 Did you mean to request changes instead of approve?

File: src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java (hunk 5 above)
Comment (AnaghHegde) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java]: forgot to update the java doc. I will work on it once i am done with my exams.

**Comment by thenaesh** [comment]
Closing due to inactivity.

**Comment by AnaghHegde** [comment]
@thenaesh I am having exams.

**Pull request closed without merging**